src/H5FDint.c

```diff
@@ -425,3 +441,285 @@ H5FD_delete(const char *filename, hid_t fapl_id)
+
+    FUNC_ENTER_STATIC_NOERR
+
+    if (H5FD_GET_DRIVER_BY_NAME == op_data->key.kind) {
+        if (0 == HDstrcmp(cls->name, op_data->key.u.name)) {
+            op_data->found_id = id;
+            ret_value         = H5_ITER_STOP;
+        } /* end if */
+    }     /* end if */
+    else {
+        HDassert(H5FD_GET_DRIVER_BY_VALUE == op_data->key.kind);
+        if (cls->value == op_data->key.u.value) {
+            op_data->found_id = id;
+            ret_value         = H5_ITER_STOP;
+        } /* end if */
+    }     /* end else */
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5FD__get_driver_cb() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD_register_driver_by_name
+ *
+ * Purpose:     Registers a new VFD as a member of the virtual file driver
+ *              class.
+ *
+ * Return:      Success:    A VFD ID which is good until the library is
+ *                          closed.
+ *
+ *              Failure:    H5I_INVALID_HID
+ *
+ *-------------------------------------------------------------------------
+ */
+hid_t
+H5FD_register_driver_by_name(const char *name, hbool_t app_ref)
+{
+    H5FD_get_driver_ud_t op_data;                     /* Callback info for driver search */
+    hid_t                ret_value = H5I_INVALID_HID; /* Return value */
+
+    FUNC_ENTER_PACKAGE
+
+    /* Set up op data for iteration */
+    op_data.key.kind   = H5FD_GET_DRIVER_BY_NAME;
+    op_data.key.u.name = name;
+    op_data.found_id   = H5I_INVALID_HID;
+
+    /* Check if driver is already registered */
+    if (H5I_iterate(H5I_VFL, H5FD__get_driver_cb, &op_data, app_ref) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_BADITER, H5I_INVALID_HID, "can't iterate over VFD ids")
+
+    /* If driver is already registered, increment ref count on ID and return ID */
+    if (op_data.found_id != H5I_INVALID_HID) {
+        if (H5I_inc_ref(op_data.found_id, app_ref) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTINC, H5I_INVALID_HID,
+                        "unable to increment ref count on VFD")
+        ret_value = op_data.found_id;
+    } /* end if */
+    else {
+        H5PL_key_t          key;
+        const H5FD_class_t *cls;
+
+        /* Try loading the driver */
+        key.vfd.kind   = H5FD_GET_DRIVER_BY_NAME;
+        key.vfd.u.name = name;
+        if (NULL == (cls = (const H5FD_class_t *)H5PL_load(H5PL_TYPE_VFD, &key)))
+            HGOTO_ERROR(H5E_VFL, H5E_CANTINIT, H5I_INVALID_HID, "unable to load VFD")
+
+        /* Register the driver we loaded */
+        /* TODO: class size calculation will not be correct with drivers
+         * that wrap the H5FD_class_t structure in their own super-structure,
+         * such as the MPI I/O VFD.
+         */
+        if ((ret_value = H5FD_register(cls, sizeof(*cls), app_ref)) < 0)
+            HGOTO_ERROR(H5E_VFL, H5E_CANTREGISTER, H5I_INVALID_HID, "unable to register VFD ID")
```

Comment:
The MPI I/O VFD wraps the H5FD_class_t structure inside its own structure, so it may be difficult to support loading it as a VFD plugin since the size calculation here will only account for the H5FD_class_t structure itself.

Reply:
This will be changing -- I don't think you will have to worry about this.